File: c++/731. My Calendar II.cpp
class MyCalendarTwo {
public:
    MyCalendarTwo() {
        
    }
    
    bool book(int start, int end) {
        for (auto &c : overlaps) {
            if (start  < c[1] && end > c[0]) {
                return false;
            }
        }

        for (auto &c : calendar) {
            if (start  < c[1] && end > c[0]) {
                overlaps.push_back({max(start, c[0]), min(end, c[1])});
            }
        }        

        calendar.push_back({start, end});
        return true;
    }

    vector<vector<int>> calendar;
    vector<vector<int>> overlaps;  // overlap记录重叠的区间
};

/**
 * Your MyCalendarTwo object will be instantiated and called as such:
 * MyCalendarTwo* obj = new MyCalendarTwo();
 * bool param_1 = obj->book(start,end);
 */

File: go/src/leetcode/731/731. My Calendar II.go
/**
@author ZhengHao Lou
Date    2022/07/19
*/
package main

type MyCalendarTwo struct {
	overlaps [][]int
	calendar [][]int
}

func Constructor() MyCalendarTwo {
	return MyCalendarTwo{}
}

func (this *MyCalendarTwo) Book(start int, end int) bool {
	for i := range this.overlaps {
		a, b := this.overlaps[i][0], this.overlaps[i][1]
		if start < b && end > a {
			return false
		}
	}

	for i := range this.calendar {
		a, b := this.calendar[i][0], this.calendar[i][1]
		if start < b && end > a {
			this.overlaps = append(this.overlaps, []int{max(start, a), min(end, b)})
		}
	}

	this.calendar = append(this.calendar, []int{start, end})
	return true
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
